src/legacy/core_plugins/embeddable_api/public/actions/action.ts

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { EuiContextMenuItemIcon } from '@elastic/eui';
+import { trackUiMetric } from '../../../ui_metric/public';
+
+import { IEmbeddable } from '../embeddables';
+
+export interface ExecuteActionContext<
+  TEmbeddable extends IEmbeddable = IEmbeddable,
+  TActionContext extends {} = {}
+> {
+  embeddable: TEmbeddable;
+  triggerContext?: TActionContext;
+}
+
+export interface ActionContext<TEmbeddable extends IEmbeddable = IEmbeddable> {
+  embeddable: TEmbeddable;
+}
+
+export abstract class Action<
+  TEmbeddable extends IEmbeddable = IEmbeddable,
+  TActionContext extends {} = {}
+> {
+  /**
+   * Determined the order when there is more than one action matched to a trigger.
+   * Higher numbers are displayed first.
+   */
+  public order: number = 0;
+  public abstract readonly type: string;
+
+  constructor(public readonly id: string) {
+    trackUiMetric('EmbeddableAPI', `actionConstructor`);
+  }
+
+  /**
+   * Optional icon that can be displayed along with the title.
+   */
+  public getIcon(context: ActionContext): EuiContextMenuItemIcon | undefined {
+    return undefined;
+  }
+
+  /**
+   * Returns a title to be displayed to the user.
+   * @param context
+   */
+  public abstract getDisplayName(context: ActionContext): string;
+
+  /**
+   * Returns a promise that resolves to true if this action is compatible given the context,
+   * otherwise resolves to false.
+   */
+  public isCompatible(context: ActionContext): Promise<boolean> {
```

Comment:
For more flexibility, though I don't think it's used asynchronously anywhere yet.  

It was originally sync and then when exploring "Create machine learning job" from ml team, that would be async because grabbing the es query used some other async function (which... maybe didn't need to be async, but it was).  

So it makes sense to me to support async operations here.

Reply:
seems that being async here will block the UI ? when user will click on the gear, he might need to wait to actually get the context menu open. same when clicking on a pie slice. And not because an action you are going to use is slow, but because some other action, which is not compatible with your embeddable even, is taking a long time to figure out if its compatible ?

sounds like a bad UX to me.

Reply:
well... if it's sync it'll block the UI and you won't see anything. If it's async, it'll block the user from seeing the menu items but we could put a spinner or something in the menu as it's loading up so in that sense the UI isn't blocked.

I get what you are saying though, we don't want to encourage users to query some other API that might take awhile to get a response back, and then this menu takes a little bit to show up.

I don't know if there is a right answer here.  more flexibility vs  potential UI slowness.... as long as implementors of actions don't abuse that it shouldn't be a problem.

Reply:
I don't feel too strongly either way, I see both cases, but let's push this to a post PR merge decision.  Then we can swap back to sync if we feel it's the right thing to do.